cocotb/triggers.py

```diff
@@ -195,6 +195,8 @@ def __init__(self, time_ps, units=None):
         See Also:
             :func:`~cocotb.utils.get_sim_steps`
         """
+        if time_ps <= 0:
+            raise TriggerException("Timer value time_ps needs to be greater than 0")
```

Comment:
Should we only do this on simulators where it's known to fail?

Reply:
The question is more like: Why would there be any sane scenario where a timer is 0 or negative?

I would consider this an error in any testcase, unless I'm overlooking something very weird.

Reply:
Zero seems semi-reasonable to me. Negative numbers I agree are nonsensical.

Reply:
> Why would there be any sane scenario where a timer is 0 or negative?

There isn't. Negative `Timer`s are obviously nonsense (time only goes one way). `Timer(0)` is undefined behavior. No standard for any language defines its meaning, and the result is that it behaves differently on every simulator. In some way it is also nonsense. `cbAfterDelay` callbacks are the first callbacks serviced in a time step. A `Timer(0)` issued later than that implies it must go back in time to the start of the time step for servicing.

Reply:
My thinking is that allowing 0 might be useful for users wanting to explore quirks in simulators that allow it.

Reply:
But what would the Time(0) semantics look like? From what @ktbarrett listed and what we found in verilator it is more than likely that it would lead to a very unpredictable behavior across simulator implementations.

And given from what I've seen over the years working with Mentor, Viviado Sim, Synopsys VCS and Cadence Incisive and Xcelium such corner cases are usually very unstable across releases. They make new optimizations without taking such stuff into account.

But as I said, I might be overlooking a reasonable use-case, but to me it looks kind of dangerous.

Reply:
The only case where it might work is in a callback in the Pre-Active region that then happens to be correctly handled by the simulator, but it's probably an edge case that's poorly handled.

There is a slightly related case for **cbAtStartOfSimTime** in the SV spec:

> Placing a callback for **cbAtStartOfSimTime** and a delay of zero during a callback for reason **cbAtStartOfSimTime** will result in another **cbAtStartOfSimTime** callback occurring during the same time slice.

That being said I can't find anything about **cbAfterDelay**

Reply:
Since there is some opposition, lets make a concession. Raise on `time < 0`, warn on `time == 0`?

Reply:
Given the literal hell I went through at Intel with VCS and such specifics of DPI/VPI: I would not rely on anything that touches those corners of the spec 😆 
It is almost certainly incorrectly interpreted or implemented or sacrificed for the sake of optimization.

Reply:
> Since there is some opposition, lets make a concession. Raise on `time < 0`, warn on `time == 0`?

Agreed, we actually have ``Timer(0)`` in at least one of our test.

Reply:
@cmarqu, does "our" mean "cocotb" or "another project"?

If that latter, that's basically what my argument is - people may have found that in the simulator they use `Timer(0)` does something useful for them, and this exception will break their code for little reason. A warning will make them aware that their code will likely fail on other simulators, but will not force them to fix it.

Reply:
I meant cocotb.

Reply:
I pushed a modified version which splits between negative and 0.
I kept the 'toplevel' if and split below that, as in the majority of the cases this check should not fire.
This saves us from having 2 toplevel 'ifs' for the sake of performance.